File: src/main/problems/java/Nr16.java
package main.problems.java;

import java.util.Arrays;

public class Nr16 {

    public static void main(String[] args) {
        Nr16 nr16 = new Nr16();
        System.out.println(nr16.threeSumClosest(
                new int[]{1, 2, 5, 10, 11},
                12
        ));
    }

    public int threeSumClosest(int[] nums, int target) {
        if (nums == null || nums.length < 2) {
            throw new IllegalArgumentException();
        }
        Arrays.sort(nums);

        int currentClosestToTarget = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < nums.length; i++) {
            int head = i + 1;
            int tail = nums.length - 1;
            while (head < tail) {
                int sum = nums[i] + nums[head] + nums[tail];
                if (sum == target) {
                    return sum;
                } else if (Math.abs(target - sum) < Math.abs(target - currentClosestToTarget)) {
                    currentClosestToTarget = sum;
                }
                if (sum > target) {
                    tail--;
                } else {
                    head++;
                }
            }
        }

        return currentClosestToTarget;
    }
}

File: src/main/problems/ruby/Nr16.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_closest(nums, target)
  nums.sort!

  closest_sum = 1 / 0.0
  nums.each_with_index do |num, idx|
    low = idx + 1
    high = nums.size - 1
    while low < high
      current_sum = num + nums[low] + nums[high]
      return current_sum if current_sum == target

      closest_sum = (target - current_sum).abs < (target - closest_sum).abs ? current_sum : closest_sum

      if current_sum > target
        high -= 1
      else
        low += 1
      end
    end
  end

  closest_sum
end
